src/ProducerEventProcessor/UploaderCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/Logging.h"
+#include "ProducerEventProcessor/UploaderCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+namespace {
+
+static constexpr const char* kAnswerString = "Test Interned String";
+static constexpr uint64_t kAnswerKey = 42;
+
+static ClientCaptureEvent CreateInternedStringCaptureEvent(uint64_t key, const std::string& str) {
+  ClientCaptureEvent event;
+  event.mutable_interned_string()->set_key(key);
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+static ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderCaptureEventCollectorTest : public testing::Test {
+ public:
+  void Start(uint64_t event_count, uint64_t produce_duration_ms, uint64_t upload_duration_ms) {
+    event_count_ = event_count;
+    produce_duration_ms_ = produce_duration_ms;
+    upload_duration_ms_ = upload_duration_ms;
+
+    wait_all_data_uploaded_duration_ms_ =
+        upload_duration_ms_ * 2 + produce_duration_ms_ * event_count_;
```

Comment:
You're trying to use sleeps for synchronization here. This kind of works but is also very brittle. It will result in flaky tests. I suggest to use Mutexes for synchronization instead. You can lock mutexes in one thread and then try to lock the same mutex in another thread. This other thread will block until the mutex is released in the first thread.

You probably need a mutex in the data producer thread that is locked for the whole time. This way you can wait in the test for the data producer to finish what it's been doing.

Reply:
Changed to use `upload_finished_mutex_` for the synchronization now. Thanks for the suggestion!